### ISSUE TITLE
Pad uuid-ncname-58 to the left with '1' characters

### ISSUE THREAD

**Issue by wrandelshofer** [doriantaylor/p5-data-uuid-ncname#2]
The draft specification 'Compact UUIDs for Constrained Grammars' specifies 3 different textual encodings: uuid-ncname-32, uuid-ncname-58, uuid-ncname-64.
https://www.ietf.org/id/draft-taylor-uuid-ncname-03.html#name-encoding-algorithm

The uuid-ncname-58 encoding is interesting to me, because it is very compact and (almost) avoids non-digit and non-letter characters. 

I propose to pad uuid-ncname-58 to the left with '1' characters instead of to the right with '_' characters.

This has the following advantages:

1. Removes the need for the '_' character in the string representation. This makes a uuid-ncname-58 visually appear more like a single word.
2. Sorting an uuid-ncname-58 lexicographically yields the same sequence like sorting the binary representation of an UUID. (This is already the case with the canonical-uuid-representation, uuid-ncname-32, and uuid-ncname-64).

Sorting is of interest for the proposed UUIDv6 format specified in https://www.ietf.org/archive/id/draft-peabody-dispatch-new-uuid-format-04.html#uuidv6

Proposed changes in
https://www.ietf.org/id/draft-taylor-uuid-ncname-03.html#name-syntax

    base58     = 6"1" 15b58char / 5"1" 16b58char /
                 4"1" 17b58char /  3"1" 18b58char /
                 2"1" 19b58char /  "1" 20b58char / 21b58char
                 ; (symbol sequence plus appropriate padding)

Proposed changes in 
https://www.ietf.org/id/draft-taylor-uuid-ncname-03.html#name-encoding-algorithm

And Base58:

1. Remove the last octet from the binary string bin, convert it to an integer and assign it to variant.
2. Shift variant to the right by 4 bits, and convert it to its value in the Base32 table.
3. Encode the remaining bin with the Base58 algorithm to get the string b58.
4. If b58 is less than 21 characters long, ~~append underscores (_) until it is.~~ **prepend ones ('1') until it is.**
5. Convert version to its value in the Base32 table.
6. Return the concatenation of version, b58, and variant.

Proposed changes in 
https://www.ietf.org/id/draft-taylor-uuid-ncname-03.html#name-decoding-algorithm

If ncname is Base58:

a. Remove the last character and decode it to an integer according to the Base32 spec; call that integer variant.
b. Shift variant four bits to the left.
c. ~~Remove all trailing underscores from the remainder of ncname.~~
d. Decode the remainder of ncname with the Base58 algorithm as bin.
e. Append the octet corresponding to the value of variant to bin.

Proposed changes in
https://www.ietf.org/id/draft-taylor-uuid-ncname-03.html#name-samples-of-uuid-ncname-58-r

Version | Base58
-- | --
0, Nil|`A111111111111111111111A`
1, Timestamp|`B6fTkmTD22KpWbDq1LuiszL`
2, DCE "Security"|`C1111KtP6Y9P3rRkvh2N1eL`
3, MD5|`D2ioV6oTr9yq6dMojd469nJ`
4, Random|`E13UZ99RxxUJC1v4dWsYtbJ`
5, SHA-1|`F1x7wEJfz9eb1TYzsrT7ZsI`

**Comment by doriantaylor** [comment]
I like this idea in principle, though can you tell me how you would disambiguate the decoded bit vector? The reason why I opted for underscores in the first place was because the underscore was still a valid identifier character but not part of the base58 alphabet, enabling it to pad to a consistent length while keeping the content unambiguous.

**Comment by wrandelshofer** [comment]
Thank you for your replying so quickly! 😀


I have made a quick-and-dirty implementation in Java in this repository:
https://github.com/wrandelshofer/UuidNCName

The [fromString() method](https://github.com/wrandelshofer/UuidNCName/blob/9b82770f2e998b0cbdf12d1c31da29fb87617895/src/main/java/ch/randelshofer/uuidncname/UuidNCName.java#L189-L196) is able to parse both variants of uuid-ncname-58 (those with trailing '_' padding and those with leading '1' padding).



The leading '1' characters decode as zero bytes.
So, when we decode a uuid-ncname-58 String, and the base58 decoder generates more than 16 bytes, then the leading extra bytes must be 0.

**Comment by wrandelshofer** [comment]
I just found out, that only uuid-canonical and my proposed uuid-ncname-58 format have the same sort order as the binary UUID representation. 

This is because in ASCII we have the following ordering of character codes: '-' < '0'-'9' < 'A'-'Z' < '_' < 'a'-'z'.

The uuid-ncname-32 and uuid-ncname-64 encode the binary UUID representation with a different character ordering, and have therfore a different sort order.

**Comment by wrandelshofer** [comment]
Is it possible to use an alternative encoding scheme for uuid-ncname-64 that preserves lexicographical order - like the one described below?

How can we disambiguate such an uuid-ncname-64-lexical from uuid-ncname-64?


http://www.faqs.org/qa/rfcc-1940.html

> 1° base64 encoding is 33% more compact than Hex encoding, but lexicographical order is not preserved. Though this property could be very useful when used inside filenames or URLs.  
> 
> A simple reordering of the proposed alternate base64 encoding table would just do the trick.
> 
> 0 : '-'
> 1..10 : '0'..'9'
> 11..36 : 'A'..'Z'
> 37 : '_'
> 38..63 : 'a'..'z'
> 
> 2° A distinct name should be suggested for this base64 variant. What about base-64 encoding ? Putting the dash in the middle makes it DNS friendly.
>

**Comment by wrandelshofer** [comment]
I have now described a syntax that allows to unamibiguosly encode/decode the UUIDs in the various formats:

https://github.com/wrandelshofer/UuidNCName/blob/20418aa8a547ad777d49b22359499e9aa27c6540/README.md

I believe it adds a lot of complexity. 
But I guess, it is not possible to change the existing UUID NCName formats anymore - they are probably already in use for quite some time (?)

What do you think?

**Comment by wrandelshofer** [comment]
I noticed that the `variant` field in UUIDv6 and UUIDv7 (actually in most UUIDs) is only 2 bits long. 

In order to preserve sort order in the lexical representation, we must not reorder bits that do not belong to the `version` and `variant` fields in these formats. 

So, instead of extracting the bits 0x00000000_0000_0000_f000_000000000000 for the `variant`, I am now extracting the bits 0x00000000_0000_0000_c000_000000000003.

The updated description is here:
https://github.com/wrandelshofer/UuidNCName/blob/baca7fe39fc24478938dda6723a277528b83d276/README.md

**Comment by doriantaylor** [comment]
I am apprehensive that mucking about with the representation of Base32/Base64 is going to be an extremely hard sell with the IETF. I also anticipate it will discourage implementers because you will no longer be able to just take an implementation off the shelf and use it; you would have to write your own codec.

Also, there is a repository [just for the spec](https://github.com/doriantaylor/draft-taylor-uuid-ncname).

**Comment by wrandelshofer** [comment]
Instead of a custom Base32 alphabet, we could use the 'radix=32' alphabet. 
This is supported in JavaScript and in Java (I haven't checked other languages).

https://tc39.es/ecma262/#sec-number.prototype.tostring
https://download.java.net/java/early_access/jdk20/docs/api/java.base/java/lang/Long.html#toString(long,int)

Oh, I missed the spec repository. Should I close this issue here and open it there?

**Comment by wrandelshofer** [comment]
I just found that [RFC 4648, Section 7,  Table 4: The "Extended Hex" Base 32 Alphabet](https://www.rfc-editor.org/rfc/rfc4648#section-7) provides a 32-bit alphabet that preserves lexicographic ordering. 
So by using this alphabet, we can solve the adoption problem of the proposed `uuid-ncname-32-lex` format.  😊

I think that `uuid-ncname-64-lex` is also not problematic, because I have found libraries that support the proposed base64-lex format exactly. For example this one: [github/deanlandolt/base64-lex](https://github.com/deanlandolt/base64-lex). 😀
